File: src/core/file_picker.m
#import <Cocoa/Cocoa.h>
#include <stdlib.h>
#include <string.h>

// 嵌入 Info.plist 到二进制，禁用 macOS AutoFill 辅助进程（macOS Tahoe 已知问题）
__attribute__((used, section("__TEXT,__info_plist")))
static const char embedded_info_plist[] =
    "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
    "<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" "
    "\"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n"
    "<plist version=\"1.0\">\n"
    "<dict>\n"
    "    <key>CFBundleIdentifier</key>\n"
    "    <string>com.zscripts.app</string>\n"
    "    <key>CFBundleName</key>\n"
    "    <string>zScripts</string>\n"
    "    <key>NSAutoFillRequiresTextContentTypeForOneTimeCodeOnMac</key>\n"
    "    <true/>\n"
    "</dict>\n"
    "</plist>\n";

// 显示文件选择对话框
char** showOpenPanel(bool allowFiles, bool allowDirectories, bool allowMultiple, const char** fileTypes, int fileTypesCount) {
    @autoreleasepool {
        NSOpenPanel* panel = [NSOpenPanel openPanel];

        // 设置选项
        [panel setCanChooseFiles:allowFiles];
        [panel setCanChooseDirectories:allowDirectories];
        [panel setAllowsMultipleSelection:allowMultiple];

        // 设置文件类型过滤
        if (allowFiles && fileTypes != NULL && fileTypesCount > 0) {
            NSMutableArray* types = [NSMutableArray arrayWithCapacity:fileTypesCount];
            for (int i = 0; i < fileTypesCount; i++) {
                [types addObject:[NSString stringWithUTF8String:fileTypes[i]]];
            }
            [panel setAllowedFileTypes:types];
        }

        // 显示对话框
        NSModalResponse response = [panel runModal];

        if (response == NSModalResponseOK) {
            NSArray* urls = [panel URLs];
            NSUInteger count = [urls count];

            if (count == 0) {
                return NULL;
            }

            // 分配路径数组（多一个位置存储数量）
            char** paths = (char**)malloc(sizeof(char*) * (count + 1));
            paths[0] = (char*)(uintptr_t)count; // 第一个元素存储数量

            for (NSUInteger i = 0; i < count; i++) {
                NSURL* url = urls[i];
                const char* path = [[url path] UTF8String];
                paths[i + 1] = strdup(path);
            }

            return paths;
        }

        return NULL;
    }
}

// 获取路径数组的数量
int getPathArrayCount(char** paths) {
    if (paths == NULL) {
        return 0;
    }
    return (int)(uintptr_t)paths[0];
}

// 释放路径数组
void freePathArray(char** paths, int count) {
    if (paths == NULL) {
        return;
    }

    for (int i = 1; i <= count; i++) {
        free(paths[i]);
    }
    free(paths);
}

File: src/core/file_picker.zig
const std = @import("std");

// Objective-C 桥接函数声明
extern "c" fn showOpenPanel(allow_files: bool, allow_directories: bool, allow_multiple: bool, file_types: [*c]const [*c]const u8, file_types_count: c_int) [*c][*c]u8;
extern "c" fn freePathArray(paths: [*c][*c]u8, count: c_int) void;
extern "c" fn getPathArrayCount(paths: [*c][*c]u8) c_int;

/// 文件选择器类型
pub const PickerType = enum {
    files, // 只选择文件
    directories, // 只选择文件夹
};

/// 文件选择器结果
pub const PickerResult = struct {
    paths: [][]const u8,
    allocator: std.mem.Allocator,

    pub fn deinit(self: *PickerResult) void {
        for (self.paths) |path| {
            self.allocator.free(path);
        }
        self.allocator.free(self.paths);
    }
};

/// 显示文件选择对话框
pub fn showFilePicker(allocator: std.mem.Allocator, picker_type: PickerType) !?PickerResult {
    const allow_files = picker_type == .files;
    const allow_directories = picker_type == .directories;
    const allow_multiple = true;

    // 文件类型过滤（只用于文件选择）
    const file_types = [_][*c]const u8{ "py", "sh" };
    const file_types_ptr = if (allow_files) @as([*c]const [*c]const u8, &file_types) else null;
    const file_types_count = if (allow_files) @as(c_int, 2) else @as(c_int, 0);

    // 调用 Objective-C 函数
    const paths_ptr = showOpenPanel(
        allow_files,
        allow_directories,
        allow_multiple,
        file_types_ptr,
        file_types_count,
    );

    if (paths_ptr == null) {
        return null; // 用户取消
    }

    // 获取路径数量
    const count = getPathArrayCount(paths_ptr);
    if (count == 0) {
        return null;
    }

    // 复制路径到 Zig 管理的内存
    var paths = try allocator.alloc([]const u8, @intCast(count));
    errdefer allocator.free(paths);

    for (0..@intCast(count)) |i| {
        const c_path = paths_ptr[i + 1]; // 实际路径从索引1开始，索引0存储的是数量
        const path_len = std.mem.len(c_path);
        paths[i] = try allocator.dupe(u8, c_path[0..path_len]);
    }

    // 释放 C 分配的内存
    freePathArray(paths_ptr, count);

    return PickerResult{
        .paths = paths,
        .allocator = allocator,
    };
}
